### DISCUSSION TITLE
`updated` lifecycle hook dont work?

### DISCUSSION THREAD

**Discussion by a1vin1au** [livewire/livewire#6400] [1. Help] [answered]
### Volt Version

1.0.0-beta.5

### Laravel Version

10.19.0

### PHP Version

8.1.22

### Description

I want to use updated function to watch count for store count value in session, but this function not working for me. Anyone have idea about this? Or I doing wrong way?

### Steps To Reproduce

```blade
<?php 
use function Livewire\Volt\{state, updated};

state(['count' => session('count', 0)]);
updated(['count' => fn () => session()->put('count', $this->count)]);

$addOne = fn () => $this->count++;
$minusOne = fn () => $this->count--;
$resetCounter = fn () => $this->count = 0;
?>

<div>
    <h1>{{ $count }}</h1>
    <button wire:click="addOne">+1</button>
    <button wire:click="minusOne">-1</button>
    <button wire:click="resetCounter">Reset</button>
</div>
```

**Comment by nunomaduro** [comment]
@a1vin1au Can you check if you are facing the same issue with regular Livewire 3?

**Comment by a1vin1au** [comment]
@nunomaduro Tried the following two scenarios, both are not working. Well, looks that is Livewire 3's bug.

```blade
<?php
use Livewire\Volt\Component;

new Class extends Component {
    public $count = 0;

    function __construct()
    {
        parent::__construct();
        $this->count = session('count', 0);
    }

    public function addOne()
    {
        $this->count++;
    }

    public function minusOne()
    {
        $this->count--;
    }

    public function resetCounter()
    {
        $this->count = 0;
    }

    public function updated($property)
    {
        if ($property === 'count') {
            session(['count' => $this->count]);
        }
    }
}
?>

<div>
    <h1>{{ $count }}</h1>
    <button wire:click="addOne">+1</button>
    <button wire:click="minusOne">-1</button>
    <button wire:click="resetCounter">Reset</button>
</div>
```

```php
<?php

namespace App\Livewire;

use Livewire\Component;

class Counter extends Component
{
    public $count = 0;

    function __construct()
    {
        $this->count = session('count', 0);
    }

    public function addOne()
    {
        $this->count++;
    }

    public function minusOne()
    {
        $this->count--;
    }

    public function resetCounter()
    {
        $this->count = 0;
    }

    public function updated($property)
    {
        if ($property === 'count') {
            session(['count' => $this->count]);
        }
    }

    public function render()
    {
        return view('livewire.counter');
    }
}
```

**Comment by a1vin1au** [comment]
I also tested all lifecycle hooks, seems none of the functions are executed

**Comment by satoved** [comment]
@a1vin1au you must not use __construct in Livewire components. Use mount() instead.

Also try a simpler version that doesn't use session to reproduce a bug.

**Comment by a1vin1au** [comment]
I checked the log that only `mount`, `boot`, `hydrate` functions are executed, while `updating`, `updated`, `rendering`, `rendered` functions are never executed.

```php
<?php

namespace App\Livewire;

use Illuminate\Support\Facades\Log;
use Livewire\Component;

class Counter extends Component
{
    public $count = 0;

    public function addOne()
    {
        $this->count++;
    }

    public function minusOne()
    {
        $this->count--;
    }

    public function resetCounter()
    {
        $this->count = 0;
    }

    public function mount()
    {
        $this->count = session('count', 0);
        Log::debug('mount');
    }

    public function hydrate()
    {
        Log::debug('hydrate');
    }

    public function boot()
    {
        Log::debug('boot');
    }

    public function updating($property, $value)
    {
        Log::debug('updating');
    }

    public function updated($property)
    {
        Log::debug('updated');
    }

    public function rendering($view, $data)
    {
        Log::debug('rendering');
    }

    public function rendered($view, $html)
    {
        Log::debug('rendered');
    }

    public function render()
    {
        return view('livewire.counter');
    }
}
```

laravel.log:
```log
[2023-08-18 08:39:46] local.DEBUG: boot  
[2023-08-18 08:39:46] local.DEBUG: mount  
[2023-08-18 08:39:49] local.DEBUG: boot  
[2023-08-18 08:39:49] local.DEBUG: hydrate  
[2023-08-18 08:39:51] local.DEBUG: boot  
[2023-08-18 08:39:51] local.DEBUG: hydrate  
[2023-08-18 08:39:53] local.DEBUG: boot  
[2023-08-18 08:39:53] local.DEBUG: hydrate  
[2023-08-18 08:39:56] local.DEBUG: boot  
[2023-08-18 08:39:56] local.DEBUG: hydrate  
```

**Comment by joshhanley** [comment]
This issue has been converted to a discussion as we are now using discussions to track bugs and ideas.
See this announcement for more details https://github.com/livewire/livewire/discussions/3147

**Comment by joshhanley** [comment, marked as answer]
@a1vin1au the updating/updated hooks only fire when a property is updated on the front end. But you are using action methods through `wire:click` so they won't fire.

If you want to test updating/updated hooks, add a `<input type="text" wire:model.live="count" />` and change the input, and the updated hook should fire.

Hope this helps!

**Reply by chadlockett** [reply]
I hope this will add some value to anyone else who might be experiencing the issue where the lifecycle events do not seem to be firing. In Livewire 3, make sure you only have one top-level HTML element in your component blade. I was experiencing an issue where updating() wasn't being called, and it was caused by having more than one root html element in my component file.

**Reply by BVFI** [reply]
Another addition to your addition. If you have one top level HTML element, but have it wrapped with ```@if/@endif``` it will not work as well and behave very weirdly. Issue is best described as a feature request for [Laravel Idea](https://github.com/laravel-idea/plugin/issues/1014)

**Comment by a1vin1au** [comment]
@joshhanley ok understood now, so if I want to watch `$count` changes but don't use `wire:model.live` attribute, I need to use `dehydrate` hook to catch it, correct?

**Comment by chrisrhymes** [comment]
I am experiencing a similar issue with Livewire v3.0.5 where I am using wire:click to push a new item to a collection. This then re-renders the frontend with the new item, but this no longer triggers the 'updated' hook like it used to in v2. 

Just wanted to check this was desired behaviour? 

```html
<button class="button" wire:click="addContact">Add Contact</button>
```

```php
    public function addContact()
    {
        $this->contacts->push(new Contact());
    }
```

**Reply by lrljoe** [reply]
See Josh's comment above!  The updated hook **will not** fire from a wire:click call.  

What are you doing that you want to call the updated hook, when you're already in the backend with your wire:click?
